core/trino-main/src/main/java/io/trino/operator/MergeOperator.java

```diff
@@ -147,6 +148,11 @@ public MergeOperator(
         this.comparator = requireNonNull(comparator, "comparator is null");
         this.outputChannels = requireNonNull(outputChannels, "outputChannels is null");
         this.outputTypes = requireNonNull(outputTypes, "outputTypes is null");
+
+        LocalMemoryContext memoryContext = operatorContext.newLocalUserMemoryContext(MergeOperator.class.getSimpleName());
+        // memory footprint of deserializer does not change over time
+        memoryContext.setBytes(deserializer.getRetainedSizeInBytes());
+        closer.register(memoryContext::close);
```

Comment:
why so? We call close() on `finish`. And memory is still not reclaimed yet.

Also you do not use this pattern in `ExchangeOperator` why a difference?

Reply:
Great catch.

I honestly don't know why are we calling `close` in `finish`.

`finish` indicates that no more pages will be added to operator via `addInput`, and it is meaningless for source operators (such as `ExchangeOperator`, `MergeOperator`) and it should not get called (and from what I understand it doesn't). 

Regardless, today both of these operators simply call `close` on `finish` what effectively means operator is no longer usable. 

Perhaps we should follow up with an another PR to resolve this confusion.